### PR TITLE
[VENTUS][fix] Fix framelowering and calculation method of stack offset

### PR DESCRIPTION
**Pull request by wangqinfan**
1. Add VMV_V_X in emitEpilogue.
2. Change all the positive numbers added by TP to negative numbers(in LowerCall).
3. Fix the LowerCall function to generate correct store instruction transferring the function parameters.
4. Fix hasReservedCallFrame function to return false.
5. Align the convention between caller and callee in the case of passing parameters by stack.
6. Change the stack offset calculation method of TP.
7. Unify the calculation of TP stack and SP stack offset.
8. Node that needing to manually modify the calculation of sp offset in the workitem.S. Since the growth direction of the stack is different from that of the traditional RISCV, it is now stipulated that for both the SP stack and the TP stack, the data is stored where the stack pointer is not offset.
9. There is a SPAdj check in eliminateFrameIndex function. but we don't need this value at all so that adding a getSPAdjust function to return zero.
10. V33 is a wrong value when parameters pushed to TP stack so there must be a MV instruction to refresh V33 after ADJCALLSTACKDOWN.

# **DO NOT FILE A PULL REQUEST**

This repository does not accept pull requests. Please follow http://llvm.org/docs/Contributing.html#how-to-submit-a-patch for contribution to LLVM.

# **DO NOT FILE A PULL REQUEST**

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp [hunk 1]
```diff
@@ -496,27 +497,38 @@ void RISCVFrameLowering::emitEpilogue(MachineFunction &MF,
...
-      StackSize +=  Size;
+      Align Alignment =
+          MFI.getObjectAlign(I).value() <= 4 ? Align(4) : MFI.getObjectAlign(I);
+      uint64_t AlignedSize = alignTo(MFI.getObjectSize(I), Alignment);
+      StackSize += AlignedSize;
     }
   }
-  return StackSize;
+
+  if (FI < 0)
```
Comment (zhoujingya) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:528]: Add comment，why here need to check `FI` < 0,  need take varidic arguments into consideration

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp [hunk 2]
```diff
@@ -623,7 +627,7 @@ static unsigned estimateFunctionSizeInBytes(const MachineFunction &MF,
 // by the frame pointer.
 // Let eliminateCallFramePseudoInstr preserve stack space for it.
 bool RISCVFrameLowering::hasReservedCallFrame(const MachineFunction &MF) const {
-  return !MF.getFrameInfo().hasVarSizedObjects();
```
Comment (oywd) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:626]: 这里为什么直接改为false？是因为ventus没有用fp指针，所以这里恒为false么？

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp [hunk 3]
```diff
@@ -698,20 +715,21 @@ RISCVFrameLowering::getFirstSPAdjustAmount(const MachineFunction &MF) const {
...
-
-  for(int I = MFI.getObjectIndexBegin(); I != MFI.getObjectIndexEnd(); I++) {
+  Align Alignment = Align(4);
+  for(int I = 0; I != MFI.getObjectIndexEnd(); I++) {
     if(static_cast<unsigned>(MFI.getStackID(I)) == ID) {
-      // Need to consider the alignment for different frame index
-      uint64_t Size = ((MFI.getObjectSize(I) + 3) >> 2) * 4;
-      StackSize += Size;
+      // FIXME: this code logic maybe not that correct?
+      StackSize += ((MFI.getObjectSize(I) + 3) >> 2) * 4;
```
Comment (oywd) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:726]: 这里应该是按照4 align，是否以后还会存在按照8 align的情况呢？

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp (hunk 2 above)
Comment (Mifuns) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:626]: 是的，所以遇到有SATCKADJUST伪指令的地方都需要处理。

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp (hunk 3 above)
Comment (Mifuns) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:726]: 这里是最小4对齐，如果更大的话后面还会根据ObjectAlign再对齐一次：
Alignment = std::max(MFI.getObjectAlign(I), Alignment);
直接用alignto应该也是可以的。

Comment (Mifuns) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp]: 这里的修改已提交。

**Comment by zhoujingya** [comment]
LGTM

**Review by zhoujingya — APPROVED** [review]

**Pull request merged**